crates/api_crud/src/user/create.rs

```diff
@@ -238,3 +227,319 @@ pub async fn register(
+        let oauth_account_form =
+          OAuthAccountInsertForm::new(user_view.local_user.id, oauth_provider.id, oauth_user_id);
+
+        OAuthAccount::create(&mut context.pool(), &oauth_account_form)
+          .await
+          .ok()
+          .ok_or(LemmyErrorType::OauthLoginFailed)?;
+
+        local_user = user_view.local_user.clone();
+
+        check_user_valid(&user_view.person)?;
+        check_registration_application(&user_view, &site_view.local_site, &mut context.pool())
+          .await?;
+        check_email_verified(&user_view, &site_view)?;
+      } else {
+        // email already registered
+        return Err(LemmyErrorType::EmailAlreadyExists)?;
+      }
+    } else {
+      // No user was found by email => Register as new user
+
+      // Extract the OAUTH name claim from the returned user_info
+      let user_name: String;
+      if let Some(user_name_value) = user_info.get(oauth_provider.name_claim) {
+        user_name = serde_json::from_value::<String>(user_name_value.clone())
+          .ok()
+          .ok_or(LemmyErrorType::OauthLoginFailed)?;
+      } else {
+        return Err(LemmyErrorType::OauthLoginFailed)?;
+      }
+
+      // remove spaces from username
+      let username = str::replace(&user_name, " ", "_");
+      let username = username + &rand::thread_rng().gen_range(0..999).to_string();
+
+      // check for slurs
+      let slur_regex = local_site_to_slur_regex(&local_site);
+      check_slurs(&username, &slur_regex)
+        .ok()
+        .ok_or(LemmyErrorType::OauthLoginFailed)?;
+
+      let require_registration_application =
+        local_site.registration_mode == RegistrationMode::RequireApplication;
+
+      if local_site.site_setup && require_registration_application && data.answer.is_none() {
+        Err(LemmyErrorType::RegistrationApplicationAnswerRequired)?
+      }
+
+      // We have to create a person, a local_user, and an oauth_account
+      person = create_person(username, &local_site, site_view.site.instance_id, &context).await?;
+
+      // Show nsfw content if param is true, or if content_warning exists
+      let show_nsfw = data
+        .show_nsfw
+        .unwrap_or(site_view.site.content_warning.is_some());
+
+      // Create the local user
+      let local_user_form = LocalUserInsertForm::builder()
+        .person_id(person.id)
+        .email(Some(str::to_lowercase(&email)))
+        .password_encrypted(None)
+        .show_nsfw(Some(show_nsfw))
+        .accepted_application(Some(!require_registration_application))
+        .email_verified(Some(oauth_provider.auto_verify_email))
+        .default_listing_type(Some(local_site.default_post_listing_type))
+        // If its the initial site setup, they are an admin
+        .admin(Some(!local_site.site_setup))
+        .build();
+
+      local_user = LocalUser::create(&mut context.pool(), &local_user_form, vec![])
+        .await
+        .ok()
+        .ok_or(LemmyErrorType::OauthLoginFailed)?;
```

Comment:
Can probably be moved into `create_person()` with some extra params and returning `(Person, LocalUser)`.